Control protocol transport layer.

bmoneke opened the pull request:
First version of the control protocol transport layer.
It is a up to date state of the current discussions regarding the control protocol in different issues.
The mermaid diagrams can be seen in #39 (in the right order)
A first rendering (if it succeeds) is in https://leco-laboratory-experiment-control-protocol--38.org.readthedocs.build/en/38/control_protocol.html

- Please use different issues for content discussion.
- Please do not remark on language, until it is "ready for review", in this first stage, we have to rig down the protocol itself.


I hope I made everything right with the formatting.

Things to determine regarding content:
- [x] How do we address the Coordinator, see #27
- [x] How do we handle heartbeats? #40 Postponed.
- [x] How to do the coordination of Coordinators (global address book)? see #22, my proposal: https://github.com/pymeasure/leco-protocol/issues/22#issuecomment-1414403570 

What are your ideas/suggestions regarding formatting:
- [x] Should we link more to the definitions?

TODO
- [x] Add link to nodes, DMT etc. after #24

bklebel commented on control_protocol.md:
```diff
@@ -0,0 +1,232 @@
+
+The receiver of a message may be specified without the Node ID, if the receiver lives in the same Node.
+
+:::{note}
+How to address the Coordinator? I used "no Recipient".
+Or a fixed name?
+:::
```
I am not sure I understand, this is now about what to do when I want to talk to the Coordinator itself? maybe repeat the Node ID? CA wants to talk directly to Co1 (for whatever reason): `| Co1.Co1 | Co1.CA | other frames  ` - when Co1 tries to look up Co1 it finds itself in its own address book

bmoneke replied:
I added `COORDINATOR` as a placeholder address, such that it is easy to replace all occurances at once.

bklebel commented on control_protocol.md:
```diff
@@ -0,0 +1,232 @@
+Every message received counts as a heartbeat.
+
+:::{note}
+TBD: Respond to every non empty message with an empty one?
```
Sounds good to me, pending discussion in #40

bilderbuchi replied:
I lament the loss of symmetry in messages, but if there's no other way to track message receipt in zmq, I guess it cannot be avoided. Deeper discussion in #40.

bklebel commented on control_protocol.md:
```diff
@@ -0,0 +1,232 @@
+    Code == "NodeID.Recipient|Co1.CA|Content" ==> NS
+    NS -- "is None" --> Local
+    NS{NodeID} -- "== Co1"--> Local
+    Local{Recipient<br> is None} -- "yes" --> Self([Message for Co1<br> itself])
```
```suggestion
    Local{Recipient<br> is Co1} -- "yes" --> Self([Message for Co1<br> itself])
```

bmoneke replied:
I added the COORDINATOR placeholder.

bklebel reviewed (approved):
Thanks for compiling this together! I almost fully agree with the content of the protocol as you compiled it here. The only suggestion which I made here now also appears in _this comment [here](https://github.com/pymeasure/leco-protocol/issues/27#issuecomment-1414016224) (#27)_

bmoneke commented:
- I added a COORDINATOR placeholder for the final name of the Coordinator.
- I expanded the Coordinator flow chart to include messages from a different Coordinator, and to include Error messages.
- I renamed the commands from (DIS)CONNECT to SIGNIN/OUT, in order to differentiate better from the socket connection.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+The {ref]}`message-layer` is the common language to understand commands, thus creating a remote procedure call.
+
+
+(transport-layer)=
```
These labels seem superfluous? Myst Parser should already autocreate the appropriate anchors (see e.g. the usage in the glossary), here even with the same anchor text.

bmoneke replied:
Does it replace spaces with colons?

bilderbuchi replied:
toLowercase, spaces to dashes, remove punctuation and other stuff like parens, slashes. Most complex example is https://github.com/pymeasure/leco-protocol/blame/8a2482956554d1bac07f30521e37d60c2df41da1/glossary.md#L40

bilderbuchi replied:
Docs for how these "slugs" are arrived at here: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#anchor-slug-structure

bilderbuchi replied:
Do you want to remove that label? Or what was your conclusion?

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+
+The following diagram shows this example communication with two Components:
+:::{mermaid}
+graph TD
```
Conceptually, a ROUTER socket has one "send" and one "recv" node (method), so I find the display here a bit confusing. 
Maybe it would be better to split into two separate graphs, first showing the flow for CA, then for CB? But then, they will be ~identical, so we don't learn that much.
Maybe this would be better as a sequence diagram with participants CA, CB, send, recv, Coordinator, and you show the flow how messages exchanged with CA and CB flow through send/recv differently? Do you know what I mean? If not I can sketch it.

bmoneke replied:
I think I know, what you mean:
CA to Co1 recv. Co1 send to CO2 recv etc.?

bmoneke replied:
Looking at the diagram, I understand you.
Is that, what you intend:
~~~mermaid
sequenceDiagram
    participant Code
    participant ROUTER as ROUTER socket
    Note over Code, ROUTER: Coordinator
    CA ->> ROUTER: "Request A"
    ROUTER ->> Code: "IA|Request A"
    Code ->> ROUTER: "IA|Reply A"
    ROUTER ->> CA: "Reply A"
    CB ->> ROUTER: "Request B"
    ROUTER ->> Code: "IB|Request B"
    Code ->> ROUTER: "IB|Reply B"
    ROUTER ->> CB: "Reply B"
~~~

bilderbuchi replied:
Exactly!

bklebel replied:
I am not sure about calling it "Coordinator internals", since already the ROUTER socket is an (internal) part of the Coordinator, although it does make one connection to the outside. I am not perfectly sure about the elegance of calling it "Code", but I think it is clearly understandable, and that is the main goal of it, I think, so for me it is fine.

bilderbuchi replied:
Followong that reasoning, the Router Socket is also "Code", though, isn't it?

bmoneke replied:
What about "Message handling"?

bilderbuchi replied:
Sounds good to me!

bklebel replied:
sounds good to me too. 

For reference and explanation, I made the distinction here between the direct _application "Code"_, controlled directly by somebody implementing this protocol, and the _zmq socket_, which although undoubtedly "lives" in some sense in "code", (not the application code implementing the protocol) is not controlled/touched in any way by a user who implements the protocol, treated as a black box outside. Into this black box, the "Code" stuffs whatever it is written to stuff, and gets out whatever the zmq socket will provide.

bilderbuchi replied:
(btw Benjamin, have you seen our Matrix room? Want to joint? Could be a good way to get a video call set up one weekend https://matrix.to/#/#pymeasure-ecp:matrix.org)

bklebel replied:
thanks for the reminder!

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+Also every Node ID has to be unique in the Network.
+As each Component belongs to exactly one Coordinator, it is fully identified by the combination of Node ID and Component ID., which is globally unique.
+This _full ID_ is the composition of Node ID, ".", and Component ID.
+For example `Co1.CA` is the full ID of the Component `CA` in the Node `Co1`.
```
It's great that you tie what we called "namespace" to Node ID -- it reuses the same term we already established (resulting in less complexity)! 
This also enables an improvement to bring more clarity to the diagrams: Just as we abbreviate components with "CA", "CB", etc., we could start calling Node namespaces "N1", "N2", etc (or "NA", "NB", if you prefer) instead of "Co1"?

bmoneke replied:
Good idea, bringing the node ID further to N1...

I like to distinguish with digits/letters the types as well, so N1.CA

bilderbuchi replied:
Thanks!

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+
+
+(address-book)=
+#### Address book
```
A refinement of terms: 
We should use different terms for different things, so it's easier to get a clear picture. What above is an "ID" is here an "Address" (iiuc); further above the random zmq designator is called "identity" (the abbrevation of which is generally "ID").

Can we/should we get a cleaner separation here, e.g. by calling the user-chosen designator "address" (local and global) and the random zmq thing "ID"? Or some other combination (local "name" and global "address", vs. "ID")?

bilderbuchi replied:
A third related/overloaded term (see e.g. https://github.com/pymeasure/leco-protocol/issues/22#issuecomment-1414403570) is the "address" meaning the host:port combination that zmq sockets live at.

Ideally, to avoid confusing ourselves and others, we find/choose different (compound) words for these different concepts, and stick to them.

bmoneke replied:
I used address for host / port (TCP), except in address book 🤦‍♂️; ID for our own naming scheme; (connection) identity for the Zmq thing, as it is thir term.

Could we use "Component name" and Node name?

bmoneke replied:
Address is given by popular usage, identity is the definition of zmq. So they are both fixed.

bmoneke replied:
I'm not sure, whether "address book" is a good name, even though it just contains names, because it is similar to a normal address book.

Or should we call it Directory?

bilderbuchi replied:
"Directory" sounds good to me!

> I used address for host / port (TCP)
> (connection) identity for the Zmq thing, as it is their term.

Agreed to both.

> ID for our own naming scheme; 

I don't like how we collide because "ID" is just the abbrevation for "identity"/identification, this is bound to confuse.

> Could we use "Component name" and Node name?

We could use the parts "Namespace" and "(Component) name"; the first we have already established. The second may also be "obviously" clear, and we could probably use "Name" alone. 
How do we call the whole thing? "Path" maybe? (not 100% convinced)

bmoneke replied:
"Full name" for the combination?

Path reminds me of file system, but on the other hand, we have a similar syntax: `path = folder.file`

bilderbuchi replied:
👍 to Full name.

bklebel replied:
full name is good for me too

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    CA ->> Co1: COORDINATOR|Co1.CA|SIGNOUT
+    Co1 ->> CA: Co1.CA|Co1.COORDINATOR|ACKNOWLEDGE
+    Note right of Co1: Removes "CA" with identity "IA"<br> from address book
+    Note left of CA: Shall not send any message anymore
```
```suggestion
    Note left of CA: Shall not send any message anymore except SIGNIN
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    Note left of CB: Reads property A
+    CB ->> Co1: Co1.CA|Co1.CB| Property A has value 5.
+    Co1 ->> CA: Co1.CA|Co1.CB| Property A has value 5.
+    Note over CA,Co1: The first message could be as well:
+    CA ->> Co1: CB|Co1.CA| Give me property A.
```
Mermaid sequence diagrams can do[ "alt" ](https://mermaid.js.org/syntax/sequenceDiagram.html#alt)to incorporate alternative paths, so this could be included after the first line.

bmoneke replied:
Oh cool. Thanks

bmoneke replied:
Done

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+:::{mermaid}
+flowchart TB
+    C([Node0.CA DEALER]) == "NodeR.Recipient|Node0.CA|Content" ==> R0
+    R0[receive] == "IA|NodeR.Recipient|Node0.CA|Content" ==> CN0{Node0 == Co1}
```
"Node0 == Co1": you mean "Co1 is in Node0", right? took me a bit to parse that.

bilderbuchi replied:
Also, a logic question? How can the answer to that ever be "no"? By definition (elsewhere), Node0.CA's dealer socket may only connect to Node0's Coordinator Co1, so "Node0==Co1" _always has to be true, no?

bmoneke replied:
The scheme started with a locally connected Component.
Then I noticed, that the Router socket might accept a message from another Coordinator and I adjusted everything, except for the inconsistency, that the Dealer socket might belong to Node0's Coordinator, not to CA directly...

> Node0 == Co1": you mean "Co1 is in Node0", right? took me a bit to parse that.

I meant, that the placeholder Node0 (which could be anything) equals the Node ID "Co1". Code example:
~~~python
Node0, ID = sender.split(",")
If Node0 == "Co1":
   ... 
~~~

bmoneke replied:
How can I increase understandibility?

bilderbuchi replied:
you reworked the whole thing, I'll check it out again and give feedback on the updated text.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    CidKnown -.->|no| E2[ERROR: ID and identity do not match]==>|"IA|Node0.CA|Co1.COORDINATOR|ERROR: ID and identity do not match"| S
+    RemIdent[remove sender identity] == "NodeR.Recipient|Node0.CA|Content" ==> CNR
+    CNR -- "is None" --> Local
+    CNR{NodeR} -- "== Co1"--> Local
```
Should that rather be "== Node0"?

bmoneke replied:
No: If the recipient Node (NodeR) is the node ID of the Coordinator (Co1), independent of the sending Node ID (Node0), then it is a local message.

bilderbuchi replied:
> the node ID of the Coordinator (Co1)

Yes, but the node ID of the Coordinator would not be `Co1`, but "Node0"/N1 -- that was the point I tried to make, but I think I made a mistake and mixed up Node0 and Node1?
I think you have this fixed already?

bmoneke replied:
Yes, I made an update.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    RemIdent[remove sender identity] == "NodeR.Recipient|Node0.CA|Content" ==> CNR
+    CNR -- "is None" --> Local
+    CNR{NodeR} -- "== Co1"--> Local
+    Local{Recipient<br>==<br>COORDINATOR} -- "yes" --> Self([Message for Co1<br> itself])
```
```suggestion
    Local{Recipient<br>==<br>COORDINATOR?} -- "yes" --> Self([Message for Co1<br> itself])
```

bmoneke replied:
I added question marks to all decision points.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    Local -- "no" --> Local2a{Recipient in Address book}
+    Local2a -->|yes, with Identity IB| Local2
+    Local2[add Recipient identity IB] == "IB|NodeR.Recipient|Node0.CA|Content" ==> R1[send]
+    R1 == "NodeR.Recipient|Node0.CA|Content" ==> W1([Wire to Co1.Recipient DEALER])
```
Is there a difference between using "**Co1**.Recipient" here and "**Node0**.CA" in the other path termination? Shouldn't both prefixes be the same? (the Node ID as that makes up the first part of the address?)

bmoneke replied:
There is a difference: Co1.Recipient, is some Component connected to Co1, as the message is delivered correctly.
The other termination returns error messages to the original sender, which is Node0.CA.

bilderbuchi replied:
> Co1.Recipient, is some Component connected to Co1

Yeah, but "some Component connected to Co1" must have the same Node namespace as Co1, and _that_ should be used here. no?

bmoneke replied:
Oh, I think you meant the discrepancy between "NodeR.Recipient" and "Co1.Recipient"?

I talked about the other "Node0.CA", which is different.

Nevertheless, let's use the other comment for discussion regarding clarity.

bilderbuchi replied:
Yes. OK.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    Local2a -->|yes, with Identity IB| Local2
+    Local2[add Recipient identity IB] == "IB|NodeR.Recipient|Node0.CA|Content" ==> R1[send]
+    R1 == "NodeR.Recipient|Node0.CA|Content" ==> W1([Wire to Co1.Recipient DEALER])
+    Local2a -.->|no| E3[ERROR Recipient unknown<br>send Error to original sender] ==>|"Node0.CA|Co1.COORDINATOR|ERROR Co1.Recipient is unknown"|CNR
```
```suggestion
    Local2a -.->|no| E3[ERROR Recipient unknown<br>send Error to original sender] ==>|"NodeR.CA|Co1.COORDINATOR|ERROR Co1.Recipient is unknown"|CNR
```

bmoneke replied:
Node0 is correct, as the error message has to return to the original Node.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+    CNR -- "== connected Coordinator Co2" --> Keep
+    Keep[send to Coordinator Co2] == "NodeR.Recipient|Node0.CA|Content" ==> R2[send]
+    R2 == "NodeR.Recipient|Node0.CA|Content" ==> W2([Wire to Co2 ROUTER])
+    CNR -- "reachable via Coordinator CoX" --> Augment
```
Should we leave out the advanced routing (more than two Coordinators) for now, and just add it in a remark/todo/issue? It could help keeping focus (and the diagrams smaller :D), and for what we imagine so far I don't think multi-hop is essential?

bmoneke replied:
Ok

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,293 @@
+Thin, dotted lines indicate decision flow in case of errors
+
+:::{mermaid}
+flowchart TB
```
Outstanding graph! I imagine it must not have been easy to produce?
Unfortunately, I can hardly read that anymore in #39. :-(

bmoneke replied:
You can click on that arrow symbol and then use pan and zoom 😁

I started with successful routing of a locally connected Component.
Then I wanted to create an error handling graph (modifying the first one) and noticed, that they should be one and that I have to include a sender from a different Coordinator (sorry for the forgotten inconsistency).

www.mermaid.live helped a lot creating the diagrams.

bilderbuchi replied:
ooh, nice! I'll bookmark that.
Good tip with the arrow - Can't figure out how to zoom without also zooming my browser (wheel or Ctrl-+/-), but I can scroll and get a wide window. If that is not sufficient, we have to find another way because that will also be unreadable on the website.

bilderbuchi commented:
Very nice work! By and large I agree with the meaning I understood, I tried to only comment on high-level/general and/or some clarity issues.
Good of you to call out what you want to focus on! When this has settled I can do some editing passes..

bmoneke commented:
Thanks for your review.

Sorry for the misunderstanding in that one graph.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+
+The following flow chart shows the decision scheme and message modification in a Coordinator `Co1` of Node `N1`.
+Its full ID is `N1.Coordinator`.
+`n0`, `nR` are placeholders for sender and recipient Node ID.
```
```suggestion
`nS`, `nR` are placeholders for sender and recipient Node ID.
```
for clearer symmetry sender/receiver?

bmoneke replied:
I was reluctant to use ns...

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+Its full ID is `N1.Coordinator`.
+`n0`, `nR` are placeholders for sender and recipient Node ID.
+`recipient` is a placeholder for the recipient Component ID.
+`IA` is the connection identity of the Component `N1.CA` (if it is directly connected to `Co1`) or of its Coordinator `n0.COORDINATOR`.
```
I think what is confusing me is the mix of generality and specificity. We have a Component `CA` on a Node `N1`, how can the name of its Coordinator be anything but `N1.COORDINATOR`? I think it shows through, as you felt you should clarify in the arrow texts that `|n0.CA|` == in fact `|N1.CA|` (which further clutters up the graph) :-/

Maybe it works better if it's all numbers/letters, or all "S" and "R" designators?

bilderbuchi replied:
Also, I get that `IA` is the connection identity of `N1.CA`, but how can it _alternatively_ be the identity of `N1.CA`'s ("its") Coordinator :interrobang:

bmoneke replied:
I tried to show the path of an arbitrary message through the system, depending on different cases.

It seems, not very successful.

> Also, I get that IA is the connection identity of N1.CA, but how can it alternatively be the identity of N1.CA's ("its") Coordinator ⁉️

Oh  yeah, that should be a variable as well .

bmoneke replied:
> Maybe it works better if it's all numbers/letters, or all "S" and "R" designators?

so, no specifics?

bilderbuchi replied:
I'd say, either all specific (fixed numbers/letters), or all generic (S/R), not a mixture.  :shrug:

bmoneke replied:
Ok. I'll try it.

bilderbuchi replied:
I think there's nothing more to be done here, resolving.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+`IB` is the connection identities of `N1.Recipient`.
+Bold arrows indicate message flow, thin lines indicate decision flow.
+Thin, dotted lines indicate decision flow in case of errors.
+Placeholder values are written in lowercase, while actually known values are begin with an uppercase letter.
```
that helps!

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+- Both Components are either connected to the same Coordinator (example one), or their Coordinators are connected to each other (example two).
+
+
+The following flow chart shows the decision scheme and message modification in a Coordinator `Co1` of Node `N1`.
```
Great idea to add a preamble explanation!

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+
+:::{mermaid}
+flowchart TB
+    C1([N1.CA DEALER]) == "nr.recipient|n0.CA|Content<br>(==nr.recipient|N1.CA|Content)" ==> R0
+    C0([n0.COORDINATOR DEALER]) == "nr.recipient|n0.CA|Content" ==> R0
+    R0[receive] == "IA|nr.recipient|n0.CA|Content" ==> Cn0{n0 == N1?}
```
Maybe this can be simplified to "we are receiving a message `nr.recipient|ns.CA|Content`? Ultimately it does not matter if it comes from a Component or a Coordinator, I think. 
Maybe use `Is` as the (generic) sender identity? Does that make sense?

bmoneke replied:
Done

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+    Cn0-->|yes| Clocal{CA in <br>local address book?}
+    Clocal -->|yes| CidKnown{IA is CA's identity<br> in address book?}
+    CidKnown -->|yes| RemIdent
+    Clocal -.->|no| E1[ERROR: Sender did not sign in] ==>|"IA|n0.CA|N1.COORDINATOR|ERROR: Sender dit not sign in<br>(==IA|N1.CA|N1.COORDINATOR|ERROR...)"| S
```
"did not sign in" -> "unknown"?

bmoneke replied:
Yes

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,296 @@
+    Cnr{nr?} -- "== N1"--> Local
+    Local{recipient<br>==<br>COORDINATOR?} -- "yes" --> Self[Message for Co1<br> itself]
+    Self == "nr.recipient|n0.CA|Content<br>(==N1.COORDINATOR|n0.CA|Content)" ==> SC([Co1 Message handling])
+    Local -- "no" --> Local2a{recipient in Address book?}
```
I think the constellation "nr? is None" should trigger an error, as the namespace is missing -- we should _only_ allow that for the sign-in, and in that case the whole recipient is/could be/should be missing (discussed elsewhere).

Or am I confused about the decision to always use full "addresses"/paths, as all Components know their namespace after signin, anyway?

bilderbuchi reviewed (commented):
Some comments from studying the big diagram. Got a bit stream-of-consciousness, please take it as feedback/comment, not always suggestions.

bmoneke commented:
- [ ] The only open question remains regarding heartbeats.
- [ ] And (almost finished) name of the Coordinator ("COORDINATOR" proposed)

bmoneke commented:
I also added the SIGN IN to another Coordinator.

Except for above mentioned (small) questions, the Transport Layer is done.

So, we can improve language now as well.

bilderbuchi commented on control_protocol.md:
```diff
@@ -280,28 +280,55 @@ For the format of the Messages, see {ref}`message-layer`.
-:::{note}
-TODO
-:::
+A Coordinator `Co1`joining a network follows a few steps:
+1. It signs in to one Coordinator `Co2` of the Network.
+2. It sends a CO_TELL_ALL message to `Co2`, to tell all other Coordinators about `Co1`s address.
+3. `Co2` tells all the Coordinators signed in (`Co3`, `Co4`...) about `Co1` with a CO_NEW message.
+4. These other Coordinators (`Co3`, `Co4`...) sign in to `Co1`.
+5. All Coordinators are connected to all others.
```
Why not
```suggestion
A Coordinator `Co1`joining a network follows a few steps:
1. It signs in to one Coordinator `Co2` of the Network.
2. If successful, `Co2` sends a list of all Coordinators (and their addresses) that it knows
3. `Co1` signs in to all Coordinators on this list that it does not know yet (step 1)
4. All Coordinators are connected to all others.
```
* Same outcome
* The responsibility for connecting remains with the connecting Coordinator
* There is not arbitrary difference between `Co2` and `Co3+` w.r.t. to the connection topology, they all were connected by `Co1`'s actions.

bmoneke replied:
That requires, that every Coordinator stores the addresses of the Coordinators.
In my setup, all Coordinators know their own address and can forget the other addresses after connecting.

> No "tell-all" message type needed. 

Yes, but you need a "give me all Coordinators" message type. 

Oh, or it is part of the local directory...?

bilderbuchi replied:
> That requires, that every Coordinator stores the addresses of the Coordinators.
In my setup, all Coordinators know their own address and can forget the other addresses after connecting.

Indeed... hmmm..

> Yes, but you need a "give me all Coordinators" message type.

No, this is just part of the sign-in flow for a coordinator.

> Oh, or it is part of the local directory...?

We would need a separate mapping from Coordinator full name to Coordinator (router) address.

bklebel replied:
Here another suggestion

```suggestion
A Coordinator `Co1`joining a network follows a few steps:
1. `Co1` signs in to one Coordinator `Co2` of the Network.
2. After successful a sign in handshake, `Co1` requests a list of all other Coordinators known to `Co2` (`Co3`, `Co4`, ...). 
3. `Co1` signs in to all the Coordinators on the list of `Co2`
4. `Co1` requests the lists of the connected Coordinators from all now connected Coordinators (except from `Co2`, it already has this one)
5. `Co1` compares the lists, and notifies Coordinators which have an incomplete set of the missing Coordinators
6. Coordinators with previously incomplete sets do sign ins to the new Coordinators they have been told about, until all Coordinators are connected to all others
```
In this case, if some Coordinator somehow drops out, those lists come out of sync, at every new sign in of a new Coordinator, the whole system brings itself back into sync. 
The only new message type would be "hey, you are missing a few more connections". 
For clarity, I separated the request for the list of connected Coordinators from the initial `SIGNIN`, although that could be part of the `SIGNIN` flow. 

> We would need a separate mapping from Coordinator full name to Coordinator (router) address.

Remember that we have a separate DEALER for every other Coordinator's connected ROUTER socket, yes, we need (in the implementation) to keep track of which DEALER socket belongs to which remote Coordinator's ROUTER address, but we need to do that anyways.

bmoneke replied:
Let us discuss in #44

bilderbuchi replied:
Sounds good, let us just drop a TODO annotation here and move on, so that we don't delay the PR more.

bilderbuchi commented on control_protocol.md:
```diff
@@ -280,28 +280,55 @@ For the format of the Messages, see {ref}`message-layer`.
+    Note over r1,d2: Sign in between two Coordinators
+    Note right of r1: shall connect<br>to address2
+    activate d1
+    Note left of d1: created with<br> name "temp-NS"
```
why do we need this "temp-NS" in the first message? 
Why doesn't N2 respond with its name during sign-in, at which point that gets stored?

bmoneke replied:
At that initial moment, N1 does not know the Node name of N2, so you need some temporary name. N1 only knows the address of N2. 

In order to know, which Dealer socket connects to which Coordinator, we need to send some identification via round trip.
Basically we could use the conversation Id, as well, but that is not yet defined.

bmoneke replied:
Another option, which breaks the dealer to router rule:

If a Coordinator N2 receives a Co Sign in message, it responds twice: once from its router to the dealer (reverse direction), such that N1 might identify the dealer socket as connected to N2; and once from its own dealer to the other router, such that N1 can store the connection identity of N2's dealer.

bmoneke replied:
> why do we need this "temp-NS" in the first message?

Maybe that helps: we send via the Dealer and receive via the Router, so in order to know, which Dealer socket connects to wich Coordinator, we have to send some piece of information (for example that temporary dealer name) via the Dealer on a round trip, we can correlate the dealer socket to the responding (received via router) Coordinator.

bmoneke replied:
Let us discuss in #44

bilderbuchi commented on control_protocol.md:
```diff
@@ -280,28 +280,55 @@ For the format of the Messages, see {ref}`message-layer`.
+        Note left of d1: name changed<br>from "temp-NS"<br>to "N2"
+        d2->>r1: GET local directory
+    end
+    d2->>r1: Here is my<br>local directory
```
wait, isn't that the reply to the 1st parallel request above? "GET local directory"?

bmoneke replied:
Yes, it is.

bilderbuchi replied:
should that be so removed from it? Or do you want to show the async nature of the communication?

bmoneke replied:
I wanted to show, that the "get local directory" message is not a response to the second sign in message, but follows immediately after the first sign in message.
The handling of the first sign in message (in N2) takes some time, until it responds.

So yes, the asynchronity causes this delay.

bilderbuchi replied:
OK, sounds good I think.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+(sign-in)=
+##### Initial connection
+
+After connecting to a Coordinator (`Co1`), a Component (`CA`) shall send a SIGNIN message indicating its Component Name.
+The Coordinator shall respond with an ACKNOWLEDGE, giving the Namespace and other relevant information, or with an ERROR, if the Component Name is already taken.
+After that successful handshake, the Coordinator shall store the connection identity and corresponding Component Name in its local {ref}`directory`.
+It shall also publish to the other Coordinators in the network that this Component signed in, see {ref}`Coordinator coordination<coordinator-coordination>`.
+Similarly, the Component shall store the Namespace and use it from this moment to generate its full name.
```
```suggestion
After connecting to a Coordinator (`Co1`), a Component (`CA`) shall send a SIGNIN message indicating its Component Name.
The Coordinator shall indicate success/acceptance with an ACKNOWLEDGE response, giving the Namespace and other relevant information, or reply with an ERROR, e.g. if the Component Name is already taken.
In that case, the Coordinator may indicate a suitable still available variation on the indicated Component Name.
The Component may retry SIGNIN with a different chosen name.

After a successful handshake, the Coordinator shall store the (zmq) connection identity and corresponding Component Name in its local {ref}`directory`.
It shall also notify the other Coordinators in the network that this Component signed in, see {ref}`Coordinator coordination<coordinator-coordination>`.
Similarly, the Component shall store the Namespace and use it from this moment to generate its full name.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+    CA ->> N1: V|N1.CB|CA|H|Content
+    Note right of N1: Does not know CA
+    N1 ->> CA: V|CA|N1.COORDINATOR|H|ERROR:I do not know you
+    Note left of CA: Should send a SIGNIN message
```
```suggestion
    Note left of CA: Must send a SIGNIN message before further messaging.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+##### Heartbeat
+
+We use heartbeat to know, whether a communication partner is still online.
```
```suggestion
Heartbeats are used to know whether a communication peer is still online.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+We use heartbeat to know, whether a communication partner is still online.
+
+Every message received counts as a heartbeat.
+
+A Component should and a Coordinator shall send a PING and wait some time before considering a connection dead.
+A Coordinator shall follow the {ref}`sign out routine<signing-out>` for a signed in Component considered dead.
+
```
```suggestion

After not having received a heartbeat from a peer for a chosen _heartbeat interval_, a Component should and a Coordinator shall send a PING and wait for a PONG response for 3 heartbeat intervals before considering a connection dead.
A Coordinator shall follow the {ref}`sign out routine<signing-out>` for a signed-in Component considered dead.
A Component may reduce its heartbeat interval when it receives a PING. 

```
3 heartbeat intervals should (I think) cover the case of a reconnecting Coordinator -- all reconnection activity should start at the latest after 1 interval.

bmoneke replied:
If the Coordinator is restarting, then it does not know the Component, so they have to sign in again.

If a Component reconnects, it has a new Connection identity and has to sign in again.

The only use case for PING is, that a Component did not send any message.
The answer should arrive within one heartbeat interval (at least if it is around 1 s).

> A Component may reduce its heartbeat interval when it receives a PING. 

What is your idea behind this?

bmoneke replied:
> thee component will send a PING, which the Component answers.

That confused me.

Now I got it.

bmoneke replied:
Let us discuss in #40

bmoneke replied:
I marked it as TBD, so it will be handled later with its own issue.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+##### Signing out
+
+A Component should tell a Coordinator, when it stops working, with a SIGNOUT message.
```
```suggestion
A Component should tell a Coordinator when it stops participating in the network with a SIGNOUT message.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+##### Signing out
+
+A Component should tell a Coordinator, when it stops working, with a SIGNOUT message.
+The Coordinator shall ACKNOWLEDGE the sign out and remove the Name from its directory.
```
```suggestion
The Coordinator shall ACKNOWLEDGE the sign-out and remove the Name from its directory.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+A Component should tell a Coordinator, when it stops working, with a SIGNOUT message.
+The Coordinator shall ACKNOWLEDGE the sign out and remove the Name from its directory.
+It shall also publish to the other Coordinators in the network that this Component signed out, see {ref}`Coordinator coordination<coordinator-coordination>`.
+
```
```suggestion
It shall also notify the other Coordinators in the network that this Component signed out, see {ref}`Coordinator coordination<coordinator-coordination>`.

```
"publish to" sounds like zmq PUB, I'm not sure that that is the intention.

bmoneke replied:
It was my intention, when I wrote it, as that was the state of discussion.

Now "notify" is better.

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+    CA ->> N1: V|COORDINATOR|N1.CA|H|SIGNOUT
+    participant N1 as N1.COORDINATOR
+    N1 ->> CA: V|N1.CA|N1.COORDINATOR|H|ACKNOWLEDGE
+    Note right of N1: Removes "CA" with identity "IA"<br> from directory
```
```suggestion
    Note right of N1: Removes "CA" with identity "IA"<br> from directory
    Note right of N1: Notifies other Coordinators about sign-out of "CA"
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+#### Communication with other Components
+
+The following two examples show, how a message is transferred between two components `CA`, `CB` via one or two Coordinators.
+Coordinators shall send messages from their DEALER socket to other Coordinator's ROUTER socket.
```
```suggestion
The following two examples show how a message is transferred between two components `CA`, `CB` via one or two Coordinators.
Coordinators shall send messages from their DEALER socket to other Coordinator's ROUTER socket.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+The following two examples show, how a message is transferred between two components `CA`, `CB` via one or two Coordinators.
+Coordinators shall send messages from their DEALER socket to other Coordinator's ROUTER socket.
+
+Coordinators shall hand on the message to the corresponding Coordinator or connected Component.
```
```suggestion
Coordinators shall route the message to the corresponding Coordinator or connected Component.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+    N1 ->> CA: V|N1.CA|N2.CB|H| Property A has value 5.
+:::
+
+Prerequisite of Communication between two Components are:
+- Both Components are connected to a Coordinator and {ref}`signed in<sign-in>`.
+- Both Components are either connected to the same Coordinator (example one), or their Coordinators are connected to each other (example two).
```
```suggestion
Prerequisites of Communication between two Components are:
- Both Components are connected to a Coordinator and {ref}`signed in<sign-in>`.
- Both Components are either connected to the same Coordinator (example one), or their Coordinators are connected to each other (example two).
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+#### Coordinator coordination
+
+Each Coordinator shall keep an up to date global {ref}`directory` with the names of all Components in the Network.
+For this, Coordinators shall tell each other regarding signing in and signing out Components and Coordinators.
+Coordinators shall send on request the Names of their local directory, or of their global directory, depending on the request type.
```
```suggestion
Each Coordinator shall keep an up-to-date global {ref}`directory` with the names of all Components in the Network.
For this, Coordinators shall tell each other about sign-ins and sign-outs of Components and Coordinators.
On request, Coordinators shall send the Names of their local directory, or of their global directory, depending on the request type.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+For the format of the Messages, see {ref}`message-layer`.
+
+
+##### Coordinator sign in
```
```suggestion
##### Coordinator sign-in
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+4. These other Coordinators (`Co3`, `Co4`...) sign in to `Co1`.
+5. All Coordinators are connected to all others.
+
+Two Coordinators shall follow a more thorough sign in/sign out procedure, than Components (address is for example host an port).
+The sign in might happen because of a CO_NEW message arrived or at startup.
+The sign out might happen because the Coordinator shuts down.
```
```suggestion
Two Coordinators shall follow a more thorough sign-in/sign-out procedure than Components (address is for example host and port).
The sign-in might happen because of a CO_NEW message arrived or at startup.
The sign-out might happen because the Coordinator shuts down.
```

bilderbuchi commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+##### Coordinator updates
+
+Whenever a Component signs in to or out of its Coordinator, the Coordinator shall send a note regarding this event to all the other Coordinators.
+The note shall contain the full name of the Component and the event type (sign in or out)
+The other Coordinators shall update their global directory according to this note (add or remove an entry).
```
```suggestion
Whenever a Component signs in to or out of its Coordinator, the Coordinator shall send a message regarding this event to all the other Coordinators.
The message shall contain the full name of the Component and the event type (sign in or out)
The other Coordinators shall update their global directory according to this message (add or remove an entry).
```
I don't think we need to introduce the term "note" in addition to "message". Might need a specific message type.

bilderbuchi reviewed (commented):
Language editing pass. As much as possible I made the edits, so you just have to press "add to batch" when in the Files view, if you agree with the suggestion.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+#### Configuration
+
+Each {ref}`Coordinator <components.md#coordinator>` shall offer one ROUTER socket, bound to a host name (or IP address or any address of a computer with "*") and port.
+
```
You can bind to a specific IP address of your computer (i.e. you have several network adapters) or you can bind to all IP addresses with "*". For example "tcp://*:12345" binds to all IP addresses at port 12345.

bmoneke replied:
I tried to make it more descriptive.

bilderbuchi replied:
Thanks!

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+#### Naming scheme
+
+Each Component has an individual name, given by the user, the _Component Name_.
+A Component Name must be a series of bytes, without the ASCII character "." (byte value 46).
```
It is easier for the coordinator, if it does not have to decode at all. Therefore the Coordinator splits the byte sequence at a byte with value 46 and uses the first part as namespace and the second part as Component name.

You can use ASCII encoding, but you do not have to.

> Whitespace allowed?

So, yes, whitespaces and values above 127 are allowed.

> Although here we should probably not talk about bytes at all, and just define what the permissible character range is (ASCII or UTF-8, except ".")

So, you're for using some encoding?

bilderbuchi replied:
Well, I imagined (for readability/debugging alone) to store the names as strings (`Lab.MyDevice` vs. `4c 61 62 2e 4d 79 44 65 76 69 63 65`), but of course you're right, it will be faster to compare the raw byte values to received frames, and for the computer it doesn't matter. Over the wire and I guess in the zmq frames we'll have bytes only.
However, all user-facing stuff (e.g. logs, printouts, debugging, interfaces) should use strings.

If we stick to printable ASCII as the allowed character set, the en/decoding will be quite simple. UTF-8, less so I imagine.

> So, yes, whitespaces and values above 127 are allowed.

Now I'm confused. What's the character for a code point of 128 (byte 80)? Do you mean UTF-8?
What if we stick to printable ASCII (0x20 to 0x7E)?

bmoneke replied:
> Now I'm confused. What's the character for a code point of 128 (byte 80)? Do you mean UTF-8?
What if we stick to the printable ASCII range (0x20 to 0x7E, excluding 0x2E) for the names?

I said any bytes value, therefore also whitespaces. Whether they are useful for a user is another question.

So, let us change it to: 'Names consist of ASCII characters excluding the period (".")'.
If a user want stupid names (including space, new line, carriage return...), its his problem.

> strings

Strings is not that easy. For example a string in C has to be zero terminated. If we stick to "byte sequence representing ASCII characters", everybody knows what to expect and what not.

bilderbuchi replied:
> If a user want stupid names (including space, new line, carriage return...), its his problem.

It becomes our problem as soon as someone starts shenanigans with "funny" names.

> A Component name must be a series of ASCII characters, without the character "."

In principle fine, I'd like to restrict this to printable ASCII, but this is an encapsulated/independent change -- I'll open an issue and resolve here.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+This _full name_ is the composition of Namespace, ".", and Component Name.
+For example `N1.CA` is the full name of the Component `CA` in the Node `N1`.
+
+The receiver of a message may be specified without the Namespace, if the receiver lives in the same Node.
```
I'm unsure, whether we should add that single exception to the name definition, or only in the section regarding the sign in process.

bmoneke replied:
Done manually. I left it open for the discussion.

bilderbuchi replied:
> I'm unsure, whether we should add that single exception to the name definition, or only in the section regarding the sign in process.

Well, definitely to the name definition here, otherwise the section on sign in might contradict the definition here: _Here we say "must be Full name" - in the sign in section we say "well, except during sign-in"._
We can have it in the sign in section, too, of course.

bmoneke replied:
I added the text, as suggested.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+#### Directory
+
+Each Coordinator shall have a list of the Components connected to it.
+This is its _local Directory_.
+
+The _global Directory_ is the combination of the local directories of all Coordinators in a Network.
```
Good idea, for easier distinction than "local directory" and "global directory".

bmoneke replied:
Done

bilderbuchi replied:
sorry for being not clear enough, I meant "put the term into the glossary?" >.<

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+Here the Message content is expressed in plain English, for the exact definition see {ref}`message-layer`.
+
+In the exchange of messages, only the messages over the wire are shown, the connection identity of the ROUTER socket is not shown.
```
It depends on the point of view:
The router socket has a hashtable with these identities, which indicate certain connections to DEALER sockets.
So, the identity lives in the router socket. It describes, however, the connection to the DEALER socket...

bmoneke replied:
I tried to make it more descriptive with "used by the ROUTER".

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+We indicate the separation between frames with `|`.
+An empty frame is indicated with two frame separators `||`, even at the beginning or end of a message.
+For example `||Second frame|Third frame||Fifth frame` consists of 5 frames.
+The first and fourth frames are empty frames.
```
Done manually.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+Whenever a message is sent to the ROUTER socket from a peer, it reads as a first frame that identity.
+For example if `CA` sends the message `Request A`, the ROUTER socket will read `IA|Request A`.
+That way, you can return an answer to exactly the original peer and not, for example `CB`.
+If you call the ROUTER's send command with `IA|Reply A`, the socket will send `Reply A` to the peer, whose connection is `IA`, in this case `CA`.
```
Done manually

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+Component Names must be unique in a {ref}`Node <network-structure.md#node>`, i.e. among the Components connected to a single Coordinator.
+The Coordinator itself has the Component Name `COORDINATOR` (ASCII encoded).
+
+Similarly, every node has a name, the _Namespace_.
```
Done manually.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+#### Naming scheme
+
+Each Component has an individual name, given by the user, the _Component Name_.
```
Maybe Directory and Glossary.

bmoneke replied:
Done

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+
+### Sockets and Connections
+
+We use [zmq](https://zeromq.org/) sockets for our communication. For more details see the [zmq guide](https://zguide.zeromq.org/) or [zmq API](http://api.zeromq.org/)
```
Ok, I moved it.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,359 @@
+(router-sockets)=
+#### Particularities of ROUTER sockets
+
+A small introduction to ROUTER sockets, for more details see [zmq guide chapter 3](https://zguide.zeromq.org/docs/chapter3/#Exploring-ROUTER-Sockets).
```
I made an appendix, where I put the router socket. The general introduction (with the frame separation) is in the network structure.

bilderbuchi commented on control_protocol.md:
```diff
@@ -100,34 +59,34 @@ A message consists of 4 or more frames.
+This is its _Directory_.
 
-The _global Directory_ is the combination of the local directories of all Coordinators in a Network.
+The _Glossary_ is the combination of the Directories of all Coordinators in a Network.
```
?? :confused:
Ups, what I meant with my comment is: "should 'global/local directory' get an entry in the glossary?"

bmoneke replied:
@bklebel I misunderstood bilderbuchis comment (in another place), where he wanted me to put these names in the glossary, and changed the name to glossary.

I'm going to revert to local/global directory.

bmoneke replied:
Done

bilderbuchi replied:
Thanks :relieved:

bklebel commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+
+##### Coordinator updates
+
+Whenever a Component signs in to or out of its Coordinator, the Coordinator shall send a CO_UPDATE message regarding this event to all the other Coordinators.
```
```suggestion
Whenever a Component signs in to or out from its Coordinator, the Coordinator shall send a CO_UPDATE message regarding this event to all the other Coordinators.
```

bklebel commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+
+Whenever a Component signs in to or out of its Coordinator, the Coordinator shall send a CO_UPDATE message regarding this event to all the other Coordinators.
+The message shall contain the Full name of the Component and the event type (sign in or out)
+The other Coordinators shall update their Glossary according to this message (add or remove an entry).
```
```suggestion
The other Coordinators shall update their global Directory according to this message (add or remove an entry).
```

bklebel reviewed (commented):
Thanks again for the good work! 
In this review, I now mostly focused on language and punctuation, only a few comments are about the core workings. I will need to catch up with you again :)

bmoneke commented on appendix.md:
```diff
@@ -0,0 +1,34 @@
+
+A small introduction to ROUTER sockets, for more details see [zmq guide chapter 3](https://zguide.zeromq.org/docs/chapter3/#Exploring-ROUTER-Sockets).
+
+A router socket assigns a random _identity_ to each connecting peer.
```
I disagree. There are several more socket types.

> Then, it  assigns a random _identity_ to each connecting peer.

Even if a ROUTER socket connects to some other socket, it assigns an _identity_ to that connection.

bklebel replied:
yes, of course, there are many more socket types, which typically serve special purposes. The ROUTER socket can be connected with most (if not all, not sure about PUB-SUB) socket types, however that does not mean that the other socket types are useless. IIRC, ROUTERs cannot initiate communication as long as they do not know an _identity_ to which they could send messages. If the application code knows the _identity_ from some other source, the ROUTER can start, but otherwise it cannot. 
This is all I wanted to put in there, to give a bit more context.

bmoneke replied:
ROUTER cannot connect to PUB-SUB, nor PAIR, nor RADIO, nor DISH...
Only to ROUTER, REP, REQ, DEALER.

Anyways, I tried to make it more clear in my change (coming soon).

bklebel replied:
Okay, I have to admit, RADIO-DISH and SERVER-CLIENT socket connections were now new to me, and PAIR was somehow always kind of out of my scope. I like your change!

bmoneke commented on appendix.md:
```diff
@@ -0,0 +1,34 @@
+Whenever a message is sent to the ROUTER socket from a peer, the socket prepends that identity in front of the message frames.
+For example if `CA` sends the message `Request A`, the ROUTER socket will read `IA|Request A`.
+That way, the answer to that message can be returned to the correct peer and not another one (a ROUTER socket can have many connected peers).
+If the ROUTER's send command is called with `IA|Reply A`, the socket will send `Reply A` to the peer, whose connection is `IA`, in this case `CA`.
```
We're using more passive voice, therefore, I'm going to rewrite your suggestion.

bmoneke replied:
Done manually.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+While the number of DEALER sockets thus required scales badly with the number of Connectors in a LECO Network, the scope of the protocol means that at most a few Coordinators will be involved.
+:::
+
+Messages must be sent to a Coordinator's ROUTER socket.
```
That you shall not send any Message to the DEALER socket of a Coordinator.

Is this better?
```suggestion
All messages must be sent to a Coordinator's ROUTER socket, not to one of its DEALER sockets.
```

bmoneke replied:
I tried to improve it, please have a look.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+Component names must be unique in a {ref}`Node <network-structure.md#node>`, i.e. among the Components (except other Coordinators) connected to a single Coordinator.
+A Coordinator itself must have the Component name `COORDINATOR` (ASCII encoded).
+
+Similarly, every Node must have a name, the _Namespace_.
```
Here I define _Namespace_ as the Node's name, therefore a link to the glossary will be a circulary link (as the glossary links here).

bklebel replied:
ok, that makes sense

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+sequenceDiagram
+    Note over CA,N1: Name "CA" is still free
+    participant N1 as N1.COORDINATOR
+    CA ->> N1: V|COORDINATOR|CA|H|SIGNIN
```
It was some idea, yes, but it is not decided.

It does not affect the working of the Transport Layer itself, therefore, I'd leave it as is and (maybe) retouch it, once the Message Layer is defined. (Even the type names are not totally fixed yet)

bmoneke replied:
The whole Message layer is still TBD, therefore, I do not see the necessity to add a note here.

bilderbuchi replied:
Still, dropping a note makes it easier to not miss to update this later, don't you think?

bmoneke replied:
I modified line 76 to express, that we put everything in the Content frame:
~~~
Here the Message content is expressed in plain English and placed in the Content frame, for the exact definition see {ref}`message-layer`.
~~~

Here I also add a note, that we have to determine, how we show encoded content (according to the Message Layer) in these examples.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+
+#### Directory
+
+Each Coordinator shall have a list of the Components (including other Connectors) connected to it.
```
We have to talk about that separately.

bmoneke replied:
see #44

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+
+##### Signing out
+
+A Component should tell a Coordinator when it stops participating in the network with a SIGNOUT message.
```
Solved differently.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,327 @@
+#### Communication with other Components
+
+The following two examples show how a message is transferred between two components `CA`, `CB` via one or two Coordinators.
+Coordinators shall send messages from their DEALER socket to other Coordinator's ROUTER socket.
```
I added a "Remember:"

bmoneke replied:
```suggestion
Remember: Coordinators shall send messages from their DEALER socket to other Coordinator's ROUTER socket.
```

bmoneke replied:
I removed it, as I expanded the connection scheme text.

bmoneke commented on control_protocol.md:
```diff
@@ -276,9 +235,9 @@ flowchart TB
-Each Coordinator shall keep an up-to-date global {ref}`directory` with the names of all Components in the Network.
-For this, Coordinators shall tell each other about sign-ins and sign-outs of Components and Coordinators.
-On request, Coordinators shall send the Names of their local directory, or of their global directory, depending on the request type.
+Each Coordinator shall keep an up-to-date {ref}`Glossary<directory>` with the Names of all Components in the Network.
```
Done

bmoneke commented:
I updated the Coordinator-sign-in procedure according to #44 (with a few improvements). An additional, flow chart of decisions is in #39 .

I suggest, we leave heartbeats open (a TBD not is there) and finish this PR.

The heartbeat can be added in its own PR.

bklebel reviewed (commented):
Thanks again! 
I agree with setting the heartbeats into a TBD, and get this PR through. 
Apart from minor changes I have two topics to consider: 

1. the message types, and reduction to `CO_UPDATE`, as per [this comment](https://github.com/pymeasure/leco-protocol/issues/29#issuecomment-1428145358)
2. Referring to the discussion in the matrix channel regarding separation of transport and message layer, it might be worthwhile to separate the abstract transport layer protocol (which does not depend on zmq, like Directories, sign-ins and Directory-updates between the Coordinators) a bit stronger from the zmq related definitions of ROUTER and DEALER sockets. However, I think we can go through with this PR, and as one of the next steps make this separation more clear in a new PR

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,351 @@
+##### Coordinator updates
+
+Whenever a Component signs in to or out from its Coordinator, the Coordinator shall send a CO_UPDATE message regarding this event to all the other Coordinators.
+The message shall contain the Full name of the Component and the event type (sign in or out)
```
In a small Network with little sign-ins/sign-outs, it seems fine, to send the whole local Direcetory.

I see, however, no reason to send the global Directory and then to filter according to the sending Namespace.

bklebel replied:
ok, that is true, so maybe we only send the local Directory, which we will define to also contain all remote Coordinators? Or do you want to have two separate send messages for the local and the global Directory, if one or the other is requested? also refer to my comment in #29

bmoneke replied:
What do you think, @bilderbuchi ?

bilderbuchi replied:
Hm, I don't have strong opinions on this. 
I don't mind more message types. 
I'd have gone with one "update" type, i.e. send names of a local Node's components (i.e. what the sending Coordinator is responsible for) and if they signed in or out. And have one "sending everything", which sends the whole directory (global and local). The latter will typically be used on coordinator sign-in, with the former being efficient delta updates on that. Every some "long" period, we can resync with "everything" updates.

bilderbuchi replied:
I opened #46 on the sign-in info propagation to keep #29 focused.

bklebel replied:
This is ok for me how it is now, we can merge this and continue the discussion in #29 and #46 separately

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,351 @@
+
+### Messages for Transport Layer
+
+- SIGNIN
+- SIGNOUT
+- ACKNOWLEDGE
+- ERROR
+- PING
+- CO_SIGNIN
+- CO_SIGNOUT
+- CO_TELL_ALL
+- CO_NEW
+- CO_UPDATE
```
I already did that (in the meantime) in a local commit.

bmoneke commented on control_protocol.md:
```diff
@@ -0,0 +1,351 @@
+
+    AR(['ACK N2' <br>at DEALER]) --> Known{DEALER name == 'N2'?}
+    Known -->|no| Rename[Rename DEALER socket to 'N2'] -->SLD
+    Known -->|yes| SLD([Send my local Directory])
```
Let us keep it in that plain English, abstract way. Once we have the messages, we can introduce it here.

bmoneke commented on glossary.md:
```diff
@@ -20,9 +23,16 @@ Device
     A Component which takes part in orchestrating a (i.e. LECO-controlled) measurement setup, see {ref}`components.md#director`.
 
+Directory
+    Each Coordinator maintains a local Directory with all the Components connected to it, and a global Directory with all Components in the Network, see {ref}`control_protocol.md#directory`.
```
I don't know, whether it is necessary to indicate how the Coordinator maintains the Directory.
Your change seems to imply, that the Directories differ in the type of names stored, but the difference is the scope: Components of the Node and Components of the Network.

The Coordinator can store the local Directory with Full names as well, if it wishes.
The global Directory has to have full names.

What about:
```suggestion
    Each Coordinator maintains a local Directory with all the Components connected to it (i.e. Components of the Node), and a global Directory with all Components in the whole Network, see {ref}`control_protocol.md#directory`.
```

bilderbuchi replied:
I agree with Benedikt. We have to take care to specify (observable) behaviour, and not implementation details (it's hard to separate at times). Also, we need to define the (technical) terms we use, so that we can understand  (here local vs. global directory) and use them efficiently in the spec.

bilderbuchi replied:
Benedikt, FYI, that there is a slight mismatch between your proposed definition of local directory, and what the main text says, i.e. 
> Each Coordinator shall have a list of the Components (including other Coordinators) connected to it.
This is its _local Directory_.

Other Coordinators are not "of the Node".

bilderbuchi replied:
I'm not sure though why we need separate global and local directories? Probably I have not picked up on all the conversation. 
Why don't we have one directory with every Coordinator? If an entry is local or global will be immediately clear from the Node namespace, and can be filtered as such at will.

bmoneke replied:
We should separate them at least in technical terms, as the local Directory is necessary for routing (it contains the connection identities as well).
And a Coordinator is the primary source of truth of that Directory.

The global directory of a Coordinator (which could be defined as "its copy of the union of all local Directories") is an aid for the Components in the Network.

If a Coordinator maintains only one Directory or a local version and a global one, is its one problem.

bmoneke replied:
Let us discuss it in #46 , as it is related.

bmoneke replied:
I updated the description a little bit.

bilderbuchi replied:
Thanks, sounds good. Let's resolve this and feed back in the result of #46 later.

bilderbuchi commented on control_protocol.md:
```diff
@@ -58,10 +62,10 @@ A message consists of 4 or more frames.
+This is its _local Directory_.
 
-The _Glossary_ is the combination of the Directories of all Coordinators in a Network.
+The _global Directory_ is the combination of the Directories of all Coordinators in a Network.
```
```suggestion
The _global Directory_ is the union of the Directories of all Coordinators in a Network.
```
More precise term here.

bilderbuchi replied:
However, with this definition for the global one, this basically becomes an unknowable quantity, as its "true" content is distributed over the Network, so in principle one single Coordinator could not even send it elsewhere.

bmoneke replied:
```suggestion
The _global Directory_ is a Coordinator's copy of the union of the local Directories of all Coordinators in a Network.
```

Is this better, as it adds the notion, that it is a local copy, which might not coincide perfectly with the mathematical union (unknowable).

bilderbuchi commented:
I managed to configure readthedocs builds (configured to fail on warning), and managed to activate the "list of unresolved conversations" so we have an easier time finding out how many are still open.

bilderbuchi commented:
There are some problems with some references/labels, I also see them when building locally.
> /home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:17: WARNING: undefined label: 'appendix.md#router-sockets'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:29: WARNING: undefined label: 'coordinator-sign-in'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:91: WARNING: undefined label: 'directory'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:91: WARNING: undefined label: 'coordinator-coordination'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:125: WARNING: undefined label: 'signing-out'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:135: WARNING: undefined label: 'directory'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:135: WARNING: undefined label: 'coordinator-coordination'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/control_protocol.md:241: WARNING: undefined label: 'directory'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/glossary.md:32: WARNING: undefined label: 'namespace'
/home/docs/checkouts/readthedocs.org/user_builds/leco-laboratory-experiment-control-protocol/checkouts/38/glossary.md:32: WARNING: undefined label: 'component-name'

Probably we have to check in myst-parser docs how to refer to autogenerated labels in the same file. Tomorrow.. ^.^

bmoneke commented:
I tried to fix it and somehow, sometimes the auto-slugs to not work (despite being shown with `myst-anchors -l 4 control_protocol.md`), sometimes the manual ones do not work (to the router socket in the appendix)...

bilderbuchi commented:
> sometimes the auto-slugs to not work (despite being shown with myst-anchors -l 4 control_protocol.md), 

turns out, for the autoslugs you _always_ have to give the filename too, even when it's a reference to inside the same file. So, in control_protocol.md, 
```
{ref}`message-layer`
```
does not work, but 
```
{ref}`control_protocol.md#message-layer`
```
works as expected.

This one: `<appendix.md#router-sockets>` was mixing autogenerated-slug-style (`filename#anchor`) and native ReST/Sphinx (directly refer the label, which has to be unique throughout the whole document), which did not work.

bmoneke commented:
Btw, here is the link to the generated documentation (also in the first post): https://leco-laboratory-experiment-control-protocol--38.org.readthedocs.build/en/38/control_protocol.html#coordinator-coordination

@bilderbuchi thanks for making the rendering possible and thanks for the explanation regarding the links (I looked at myst parser and was puzzled at first, we we use `{ref}...`, while the myst documentation gives the normal markdown syntax for links, until I understood it).

I removed the Coordinator's addresses and Component connection ID from the local Directory, such that the Directory is independent of the Transport Layer's underlying protocol (zmq).

That way, we push the discussion regarding Directory updates to the Message Layer (here I only request, that they notify, not whether the notification happens via a diff update or a full version).

With these changes, I hope we can finish this PR.

bmoneke commented:
In https://github.com/pymeasure/pyleco/pull/1 I have a Coordinator, where you may look at the Directory implementation (there are a couple of dictionaries needed, all in the init with a comment indicating what the key and what the value is).

bmoneke commented:
If the conversations are resolved for you, @bklebel, you may merge.

bklebel reviewed (approved):
LGTM
Sorry for the longer delay, I agree, I am going to merge this now
